src/db/pool.cr

```diff
@@ -91,10 +100,19 @@ module DB
         resource
       end
 
-      res.before_checkout
+      if res.responds_to?(:before_checkout)
+        res.before_checkout
+      end
       res
     end
 
+    def checkout(&block : T ->)
+      connection = checkout
+      yield connection
+    ensure
+      release connection if connection
```

Comment:
why the `if`?

Reply:
An exception could theoretically be raised in `checkout`, so there are no compile-time guarantees that `connection` is non-nil.

I could scope the `ensure` block only to the `yield` if we want to remove the `if`:

```crystal
def checkout(& : T ->)
  connection = checkout

  begin
    yield connection
  ensure
    release connection
  end
end
```

But I don't know if that's any clearer.

Reply:
the latter seems cleaner to me.